src/tests/ftest/nvme/enospace.py

```diff
@@ -0,0 +1,440 @@
+
+        Test Description: IO gets DER_NOSPACE when SCM is full and it release
+                          the size when container destroy with Aggregation
+                          set on time mode.
+
+        Use Case: This tests will create the pool. Set Aggregation mode to Time.
+                  Fill 75% of SCM size which will trigger the aggregation
+                  because mode set to time, next fill 75% more which will fill
+                  up NVMe. Try to fill 60% more and now SCM size will be full
+                  too. Verify last IO fails with DER_NOSPACE and SCM/NVMe
+                  pool capacity is full.
+
+        :avocado: tags=all,hw,medium,nvme,ib2,full_regression
+        :avocado: tags=der_enospace,enospc_time,enospc_time_bg
+        """
+        print(self.pool.pool_percentage_used())
+
+        # Enabled TIme mode for Aggregation.
+        self.pool.set_property("reclaim", "time")
+
+        #Run IOR to fill the pool.
+        self.run_enospace_with_bg_job()
+
+    @skipForTicket("DAOS-4846")
+    def test_enospace_time_with_fg(self):
+        """Jira ID: DAOS-4756.
+
+        Test Description: Fill up the system (time aggregation mode) and
+                          delete all containers in loop, which should release
+                          the space.
+
+        Use Case: This tests will create the pool. Set Aggregation mode to Time.
+                  Fill 75% of SCM size which will trigger the aggregation
+                  because mode set to time, next fill 75% more which will fill
+                  up NVMe. Try to fill 60% more and now SCM size will be full
+                  too. Verify last IO fails with DER_NOSPACE and SCM/NVMe
+                  pool capacity is full. Delete all the containers.
+                  Do this in loop for 10 times and verify space is released.
+
+        :avocado: tags=all,hw,medium,nvme,ib2,full_regression
+        :avocado: tags=der_enospace,enospc_time,enospc_time_fg
+        """
+        print(self.pool.pool_percentage_used())
+
+        # Enabled TIme mode for Aggregation.
+        self.pool.set_property("reclaim", "time")
+
+        #Repeat the test in loop.
+        for _loop in range(10):
+            #Run IOR to fill the pool.
+            self.run_enospace_with_bg_job()
+            #Delete all the containers
+            self.delete_all_containers()
+
+    @skipForTicket("DAOS-4846")
+    @skipForTicket("DAOS-5403")
+    def test_performance_storage_full(self):
+        """Jira ID: DAOS-4756.
+
+        Test Description: Verify IO Read performance when pool size is full.
+
+        Use Case: This tests will create the pool. Run small set of IOR as
+                  baseline.Start IOR with < 4K which will start filling SCM
+                  and trigger aggregation and start fill up NVMe.
+                  Check the IOR baseline read number and make sure it's close
+                  to the number ran prior system storage was full.

```

Comment:
When the system is full, aggregation ULT will be busy even when there is IO activities, so that'll impact the read performance, especially when there are multiple containers. Since aggregation ULT is per container, In multiple containers case, each schedule cycle will have to execute N aggregation ULTs + 1 read request (suppose you are using single synchronous IOR to generate reads, so the inflight read is always 1). You can see the performance impact will be smaller on a system with heavy workload (multiple inflight IOs), because in that case, each schedule cycle will have to execute N aggregation ULT and M read requests, the smaller N/M, the lighter performance impact.

FYI, I'm going to optimize aggregation to skip containers/objects/keys not being updated since last aggregation scan, that'll be helpful for reducing aggregation overhead in general.

Reply:
Ok, So nothing needs to be change in test case or add any specific test case?